### ISSUE TITLE
Misleading diagnostic span using trait projections in arguments

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#60980]
I came across an interesting case today where the span for a diagnostic isn't quite leading to the best location. [When compiling the example code below](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0ff311e699df06307e1256eea2508119)

```rust
struct A;

trait IntoWasmAbi {
    type Abi;
    fn into_abi(self) -> Self::Abi;
}

type T = <A as IntoWasmAbi>::Abi;

fn foo(a: T) {}
```

you get:

```
error[E0277]: the trait bound `A: IntoWasmAbi` is not satisfied
  --> src/lib.rs:10:1
   |
10 | fn foo(a: T) {}
   | ^^^^^^^^^^^^^^^ the trait `IntoWasmAbi` is not implemented for `A`
```

but it's sort of misleading because neither `A` nor `IntoWasmAbi` is mentioned in the highlight. It'd be ideal if the error would point to the `type` definition to indicat that's where the source of the error comes from.

**Comment by estebank** [comment]
The following is where the obligation could be changed from `MiscObligation` to something more specific:

https://github.com/rust-lang/rust/blob/c20654ebc318004a98303d2d5d84eee8bf44080c/src/librustc/infer/mod.rs#L1528-L1550

And here is where we could peek at that new obligation and extend it with further information.

https://github.com/rust-lang/rust/blob/c20654ebc318004a98303d2d5d84eee8bf44080c/src/librustc/traits/project.rs#L417-L455

Doing the naïve change `cause.span = tcx.def_span(tcx.associated_item(projection_ty.item_def_id).container.id())` would cause this error to output:

```
error[E0277]: the trait bound `A: IntoWasmAbi` is not satisfied
 --> file.rs:3:1
  |
3 | / trait IntoWasmAbi {
4 | |     type Abi;
5 | |     fn into_abi(self) -> Self::Abi;
6 | | }
  | |_^ the trait `IntoWasmAbi` is not implemented for `A`
```

The proper solution would require to track `type`s (which would probably need #21934 to be fixed), but would ideally look like this:

```
error[E0277]: the trait bound `A: IntoWasmAbi` is not satisfied
 --> file.rs:9:1
  |
1 |   struct A;
  |   -------- this struct doesn't implement `IntoWasmAbi`
2 | 
3 | / trait IntoWasmAbi {
4 | |     type Abi;
5 | |     fn into_abi(self) -> Self::Abi;
6 | | }
  | |_^ this trait is not implemented for `A`
7 | 
8 |   type T = <A as IntoWasmAbi>::Abi;
  |   -------------------------------- `IntoWasmAbi` expected to be implemented for `A` because of this
  | 
9 |   fn foo(a: T) {}
  |             - `IntoWasmAbi` is not implemented for `A`
```

**Comment by estebank** [comment]
@rust-lang/lang: we could assert well-formedness of `hir::ItemKind::TyAlias` early in `check_item_well_formed` (we don't today, we just skip them 👀), which would turn this case into the following more understandable output:

```
error[E0277]: the trait bound `A: IntoWasmAbi` is not satisfied
 --> file2.rs:8:10
  |
8 | type T = <A as IntoWasmAbi>::Abi;
  |          ^^^^^^^^^^^^^^^^^^^^^^^ the trait `IntoWasmAbi` is not implemented for `A`

error[E0277]: the trait bound `A: IntoWasmAbi` is not satisfied
  --> file2.rs:10:1
   |
10 | fn foo(a: T) {}
   | ^^^^^^^^^^^^^^^ the trait `IntoWasmAbi` is not implemented for `A`
```

Is this something we would consider doing, even if it is as a warning? I believe this would be effectively a breaking change (code that compiles today would stop compiling) but not "in effect" because the previously compiling code _could never be used_. This seems to me like an accident to me.
